### PR TITLE
Adding IndexMap to History Sets

### LINKED ISSUE #1256: [TASK] _indexMap for all histories
--------
Issue Description
--------

Currently, if a realization contains multiple independent indices on which sampled variables depend, a special meta key `_indexMap` is provided that determines the mapping between dimensions of the sampled variable and the names of the corresponding indices.

However, if a realization only depends on a single index (or `pivotParameter`), there is no information in the realization to indicate what the index variable is, and which variables depend on it.

For consistency of use within RAVEN and other ExternalModels, this should be fixed so the `pivotParameter` can be inquired.

----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [x] 1. Is it tagged with a type: defect or task?
- [x] 2. Is it tagged with a priority: critical, normal or minor?
- [x] 3. If it will impact requirements or requirements tests, is it tagged with requirements?
- [x] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users.
- [x] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.)

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [x] 1. If the issue is a defect, is the defect fixed?
- [x] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.)
- [x] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)?
- [x] 4. If the issue is a defect, does it impact the latest release branch? If yes, is there any issue tagged with release (create if needed)?
- [x] 5. If the issue is being closed without a pull request, has an explanation of why it is being closed been provided?

### PR DESCRIPTION
**Pull request by PaulTalbot-INL**
--------
Pull Request Description
--------
##### What issue does this change request address? (Use "#" before the issue to link it, i.e., #42.)
Closes #1256 

##### What are the significant changes in functionality due to this change request?
Adds the special keyword `_indexMap` to the `HistorySet` as used in the `DataSet`. This meta variable expresses the mapping between variables and their dependent indices (a.k.a. `pivotParameter` for a `HistorySet`).

This allows better realization comprehension in models, especially `ExternalModels`.

This doesn't change anything user facing, it is only an internal tracking.

##### Notes
This also attempts to address sporadic but common failures of the BasicStatistics print_xml test on Windows, which due to low sampling and collinearity often obtains results that differ significantly from the gold file.

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code.
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [x] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [x] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [x] 8. If an analytic test is changed/added is the the analytic documentation updated/added?
- [x] 9. If any test used as a basis for documentation examples (currently found in `raven/tests/framework/user_guide` and `raven/docs/workshop`) have been changed, the associated documentation must be reviewed and assured the text matches the example.

### CODE REVIEW

File: tests/framework/PostProcessors/BasicStatistics/print_xml.xml [hunk 1]
```diff
@@ -86,7 +88,7 @@
   <Samplers>
     <MonteCarlo name="MC_external">
       <samplerInit>
-        <limit>4</limit>
+        <limit>1000</limit>
```
Comment (PaulTalbot-INL) [inline tests/framework/PostProcessors/BasicStatistics/print_xml.xml:91]: Note this change had minimal impact on the run time, but vastly increases the reliability.

File: framework/DataObjects/DataSet.py [hunk 2]
```diff
@@ -866,14 +867,47 @@ def _checkRealizationFormat(self, rlz, indexMap=None):
...
+              if not okay:
+                break
+            else:
+              okay = False
+              break
+        if not okay:
+          # update extra/missing in case there have been changes
+          extra = mapIndices-set(self.indexes)
+          missing = set(self.indexes) - mapIndices
+          self.raiseAWarning('Realization indexes do not match expected indexes!\n',
```
Comment (alfoa) [inline framework/DataObjects/DataSet.py:907]: should not be an error?

File: framework/DataObjects/DataSet.py [hunk 3]
```diff
@@ -866,14 +867,47 @@ def _checkRealizationFormat(self, rlz, indexMap=None):
...
+        extra = mapIndices-set(self.indexes)      # extra indices not expected
+        missing = set(self.indexes) - mapIndices  # indices expected but not provided
+        if extra:
+          # perhaps someday we can collapse dimensions intelligently, but for not, this is an error state
+          okay = False # if there's extra indices listed that aren't part of the DataObject, we don't handle this yet
+        elif missing:
+          # if the variables depending on an index ONLY depend on that one index, we can infer the
+          #   structure and allow the contributing source to not explicitly provide the index map.
+          #   Should we be, though? Should this be an action of the CSV loader? a Realization class?
+          for missed in missing:
```
Comment (alfoa) [inline framework/DataObjects/DataSet.py:886]: this shouldn't be too bad in terms of speed

File: framework/DataObjects/DataSet.py (hunk 2 above)
Comment (PaulTalbot-INL) [reply framework/DataObjects/DataSet.py:907]: It does raise the error as soon as it returns; this is kind of a printout message before the error is raised.

**Comment by alfoa** [comment]
Checklist passed...

PR approved to be merged...

Merging

**Pull request merged**